x-pack/plugins/encrypted_saved_objects/server/create_migration.ts

```diff
@@ -65,19 +67,37 @@ export const getCreateMigration = (
+    const originId = encryptedDoc.originId;
+
+    // If an object is slated to be converted, it should be decrypted flexibly:
+    // * If this is an index migration:
+    //   a. If there is one or more pending migration _before_ the conversion, the object will be decrypted and re-encrypted with its
+    //      namespace in the descriptor. Then, after the conversion, the object will be decrypted with its namespace and old ID in the
+    //      descriptor and re-encrypted with its new ID (without a namespace) in the descriptor.
+    //   b. If there are no pending migrations before the conversion, then after the conversion the object will be decrypted with its
+    //      namespace and old ID in the descriptor and re-encrypted with its new ID (without a namespace) in the descriptor.
+    // * If this is *not* an index migration, then it is a single document migration. In that case, the object will be decrypted and
+    //   re-encrypted without a namespace in the descriptor.
+    // To account for these different scenarios, when this field is set, the ESO service will attempt several different permutations of
+    // the descriptor when decrypting the object.
+    const convertToMultiNamespaceType =
+      !!context.convertToMultiNamespaceTypeVersion &&
+      Semver.lte(context.migrationVersion, context.convertToMultiNamespaceTypeVersion);
+
     // If an object has been converted right before this migration function is called, it will no longer have a `namespace` field, but it
     // will have a `namespaces` field; in that case, the first/only element in that array should be used as the namespace in the descriptor
     // during decryption.
-    const convertToMultiNamespaceType =
-      context.convertToMultiNamespaceTypeVersion === context.migrationVersion;
-    const decryptDescriptorNamespace = convertToMultiNamespaceType
-      ? normalizeNamespace(encryptedDoc.namespaces?.[0]) // `namespaces` contains string values, but we need to normalize this to the namespace ID representation
-      : encryptedDoc.namespace;
+    const decryptDescriptorNamespace =
+      convertToMultiNamespaceType && encryptedDoc.namespaces?.length
+        ? normalizeNamespace(encryptedDoc.namespaces[0]) // `namespaces` contains string values, but we need to normalize this to the namespace ID representation
+        : encryptedDoc.namespace;
 
     const { id, type } = encryptedDoc;
     // These descriptors might have a `namespace` that is undefined. That is expected for multi-namespace and namespace-agnostic types.
     const decryptDescriptor = { id, type, namespace: decryptDescriptorNamespace };
-    const encryptDescriptor = { id, type, namespace: encryptedDoc.namespace };
+    const encryptDescriptor = { id, type, namespace: encryptedDoc.namespace }; // It would be preferable to rely on getDescriptorNamespace() here, but that requires the SO type registry which can only be retrieved from a promise, and this is not an async function
```

Comment:
We can't use `getDescriptorNamespace()` here without changing the ESO plugin contract.

As a result, with the current implementation: if a shareable or share-capable object (`namespaceType: 'multiple'` or `namespaceType: 'multiple-isolated'`) has both `namespaces` and `namespace` fields defined (which Should Never Happen™), no matter how the object gets decrypted, it gets re-encrypted using the `namespace` field. In practice this should never happen, but it smells bad to me.

I included unit test cases that exercise this behavior.

Reply:
Update, added enough info in the migration context to avoid this situation, see 647d3c4